src/main/java/com/google/devtools/build/lib/bazel/bzlmod/IndexRegistry.java

```diff
@@ -423,6 +439,20 @@ public Optional<ImmutableMap<Version, String>> getYankedVersions(
     }
   }
 
+  @Override
+  public boolean shouldFetchYankedVersions(ModuleKey selectedModuleKey) {
+    // If the source.json hash is known, this module has been selected during the last (successful)
+    // selection recorded in the lockfile (possibly one of two selections in case of a git merge
+    // conflict resolution). We do not fetch yanked versions for this module unless we know that it
+    // was yanked at that point but explicitly allowed via BZLMOD_ALLOW_YANKED_VERSIONS or
```

Comment:
why do we fetch yanked versions in this case? wouldn't that cause `--lockfile_mode=ERROR` to crash?

Reply:
Good catch, I'll add a test and handle this case.

I introduced this field as I would find it wrong to allow a build to succeed when a known yanked module is no longer explicitly allowed. What do you think about that?

In ERROR mode, that should probably just fail though and of course not crash. Maybe I'll just turn the crash into a user-facing error after all? ;⁠-⁠)

Reply:
did we ever have an approach where the yanked versions of each module was stored in the lockfile? We could then only refresh that list when this "new source.json" condition happens. That would also mean we correctly fail on a yanked version if the only thing that changed was `--allow_yanked_versions`.

Sorry if we switched away from that for some reason I forgot -- this whole thing has been hard to keep track of in my head.

Reply:
incidentally, that would be sort of analogous to storing the hash of `metadata.json` in the lockfile, and only refreshing it upon "new source.json". That probably simplifies the code a lot too (no need to have `useChecksum`, or pass around `yankedButAllowedModules`, etc). I find this acceptable as we're going to have a "force refresh" mode for the lockfile anyway.

Reply:
This would simplify the implementation, but I fear that it would impact usability. We would end up storing a lot of mutable, free text information about module versions that aren't even in your unpruned dep graph in the lockfile. We would also end up making the lockfile visible non-reproducible and could end up with merge conflicts in yank infos (two devs doing the same update, but yanked info for unrelated versions changed upstream in-between their resolutions). Compared to that `yankedButAllowedVersions` isn't even populated if you don't explicitly allow a yanked version.

What do you think of just failing the build (not crashing) in error mode when a previously allowed yanked version is no longer allowed? We can tell the user to run in update mode to see more details (i.e. the yanked info).

Reply:
that's a fair point. What about storing _only_ the yanked info(s) of selected versions? There isn't really any point in storing the yanked infos of unselected versions anyway, since if the user changed their deps to instead select them, we'd trigger the "new source.json" condition and refetch metadata.json.

The reason I keep harping on this point is that I found this field (yankedButAllowedVersions) very confusing (what's it used for?) and rather invasive (passed around a lot). That combo adds quite a bit to the maintenance burden IMO, despite it being just one field.

Reply:
roughly yes -- but this is stored whether or not it's allowed. And indeed `useChecksum` will need to stay -- I was thinking about the "just store the hash of source.json" approach with that, but it wouldn't work anyway because metadata.json changes could cause a checksum mismatch if it's not already cached.

Reply:
But if it's not allowed, wouldn't selection just fail? I don't see how it could end up in the lock file in that case.

Reply:
ah that's true! so it turns out to be roughly the same. Maybe it's all in the naming.

Reply:
I am now storing the yanked info in the lockfile for selected versions. I also refactored and cleaned up the logic that decides whether to fetch yanked versions or retrieve them from the lockfile.